### PR TITLE
RavenDB-9737

### PR DESCRIPTION
**Pull request by redknightlois**
2GB Memory Capped Docker Container dies after trying to free memory repeatedly because of several potential reasons. We are overly aggresive to reserve memory and overly conservative when returning it. I balanced out both directions, being more aggresive to return and less aggresive to reserve specially when operating conditions are not the gold standard.

### CODE REVIEW

File: src/Raven.Client/Http/HttpCache.cs [hunk 1]
```diff
@@ -76,51 +87,48 @@ public void Release()
...
-                }
-                catch (ObjectDisposedException)
-                {
-                    // nothing that can be done here
-                }
+#if !RELEASE
+                // Hitting this on DEBUG and/or VALIDATE and getting a higher number than 0 means we have a leak.
+                // On release we will leak, but wont crash. 
+                if (_usages > 0)
+                    throw new Exception("Detected a leak on HttpCache when running the finalizer.");
```
Comment (ayende) [inline src/Raven.Client/Http/HttpCache.cs:99]: Better exception, also add the item url so we'll have something to help us track down the cause.

File: src/Raven.Client/Http/HttpCache.cs [hunk 2]
```diff
@@ -149,70 +155,72 @@ public void SetNotFound(string url)
...
                 {
-                    sizeCleared += FreeItem(item);
+                    item.Value.ReleaseRef();
+                    sizeCleared += item.Value.Size;
                 }
+
+                sizeCleared += value.Size;
             }
 
             _isFreeSpaceRunning.Lower();
```
Comment (ayende) [inline src/Raven.Client/Http/HttpCache.cs:205]: Shouldn't this be in a try finally?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Http/HttpCache.cs (hunk 2 above)
Comment (redknightlois) [reply src/Raven.Client/Http/HttpCache.cs:205]: Yeah, it should. Correcting.

File: src/Raven.Client/Http/HttpCache.cs (hunk 1 above)
Comment (redknightlois) [reply src/Raven.Client/Http/HttpCache.cs:99]: Couldnt find any that makes sense, do you have any particular one in mind?

Comment (ayende) [reply src/Raven.Client/Http/HttpCache.cs:99]: Low memory exception? BAsically I just hate the bare exception stuff.

**Comment by redknightlois** [comment]
Rebased: Done changes in place.

**Pull request merged**